modules/caddyhttp/metrics.go

```diff
@@ -0,0 +1,215 @@
+package caddyhttp
+
+import (
+	"context"
+	"net/http"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+)
+
+var httpMetrics = struct {
+	init                   sync.Once
+	requestInFlight        *prometheus.GaugeVec
+	requestCount           *prometheus.CounterVec
+	requestErrors          *prometheus.CounterVec
+	requestDuration        *prometheus.HistogramVec
+	requestDurationSummary *prometheus.SummaryVec
+	requestSize            *prometheus.HistogramVec
+	responseSize           *prometheus.HistogramVec
+	responseLatency        *prometheus.HistogramVec
+	responseLatencySummary *prometheus.SummaryVec
+}{
+	init: sync.Once{},
+}
+
+func initHTTPMetrics() {
+	ns := "caddy"
+	sub := "http"
+
+	basicLabels := []string{"server", "handler"}
+	httpMetrics.requestInFlight = promauto.NewGaugeVec(prometheus.GaugeOpts{
+		Namespace: ns,
+		Subsystem: sub,
+		Name:      "requests_in_flight",
+		Help:      "Number of requests currently handled by this server.",
+	}, basicLabels)
+	httpMetrics.requestErrors = promauto.NewCounterVec(prometheus.CounterOpts{
+		Namespace: ns,
+		Subsystem: sub,
+		Name:      "request_errors_total",
+		Help:      "Number of requests resulting in middleware errors.",
+	}, basicLabels)
+	httpMetrics.requestCount = promauto.NewCounterVec(prometheus.CounterOpts{
+		Namespace: ns,
+		Subsystem: sub,
+		Name:      "requests_total",
+		Help:      "Counter of HTTP(S) requests made.",
+	}, basicLabels)
+
+	// TODO: allow these to be customized in the config
+	latencyBuckets := []float64{.01, .05, .1, .2, .4, 1, 3, 8, 20, 60, 120}
+	sizeBuckets := prometheus.ExponentialBuckets(256, 4, 8)
+	quantileBuckets := map[float64]float64{0.1: 0.01, 0.5: 0.01, 0.95: 0.01, 0.99: 0.001, 0.999: 0.0001}
+
+	httpLabels := []string{"server", "handler", "code", "method"}
+	httpMetrics.requestDuration = promauto.NewHistogramVec(prometheus.HistogramOpts{
+		Namespace: ns,
+		Subsystem: sub,
+		Name:      "request_duration_seconds",
+		Help:      "Histogram of round-trip request durations.",
+		Buckets:   latencyBuckets,
+	}, httpLabels)
+	httpMetrics.requestDurationSummary = promauto.NewSummaryVec(prometheus.SummaryOpts{
```

Comment:
Summary metrics are considered an obsolete instrumentation method[0]. This is especially true when you support histograms of the same data. Summary is a bit of a leftover from the early days of Prometheus before histogram functions were supported.

Let's remove these Summary metrics.

[0]: https://prometheus.io/docs/practices/histograms/

Reply:
I'd say Summaries definitely have their uses, but whenever you want to enable aggregation across instances (e.g. calculate 99th percentile latency across a cluster), you need a Histogram.

Reply:
Hmm... First I'm aware of them being considered obsolete!

Here's my perspective - summaries are useful to me for a more accurate view of "what is the actual q0.999 latency on this endpoint?", whereas I use histograms for getting alerted on latency SLO violations. I generally try to keep my number of histogram buckets limited for cardinality, clustered around the latency target, but there are circumstances where it's been very useful to have a summary. For example - if my histogram buckets max out at 30s, and I see the `+Inf` bucket growing at an alarming rate, it can be very useful to know just how bad things are - i.e. is it 35s, or is it 10mins?

All that being said, I feel like the new exemplar support in the OM format may help to answer that kind of question at some point, but I don't know enough yet about exemplars and the tooling available to consume them.

But, seeing as this is the initial metrics support, I'm totally OK with dropping summaries for now, and reconsidering later.

Reply:
The problem is, by including both, you're wasting space that could be used for more granular histogram buckets. With enough buckets, all of answers that you could get from a summary are possible with a histogram. And they can be aggregated across hosts.

This is why I highly recommend against them. People very often make the assumption that `avg(summary{quantile="0.99"})` is a valid expression and is the same as `histogram_quantile(0.99, sum by (le) (rate(histogram[1m])))`. But in reality the summary math is invalid and will give you a false impression that you're looking at good data.

Reply:
@SuperQ right, you can't avg or sum a summary (though min/max are mostly legit, as long as you understand what you're looking at). But `histogram_quantile` can also be very misleading if the bucket spread isn't ideal.

> With enough buckets, all of answers that you could get from a summary are possible with a histogram.

In my experience I'd need a lot more buckets to get an accurate-enough summary from `histogram_quantile` than I use up by simply recording a summary in the first place 😉.

But you do make a good point about assumptions - summaries can be misleading if handled incorrectly by someone inexperienced. And we want to make this accessible to users inexperienced with Prometheus.

I'll remove them for now.